File: KataFinancialTracker.f90
program financial_tracker
    implicit none
    
    type transaction
        character(len=50) :: description
        real :: amount
        character(len=10) :: type
    end type transaction

    integer, parameter :: max_trans = 100
    type(transaction) :: transactions(max_trans)
    integer :: transaction_count = 0
    real :: balance = 0.0
    integer :: choice, i, io
    character(len=50) :: desc
    real :: amt
    character(len=10) :: input

    print *, "Simple Financial Tracker"
    print *, "------------------------"
    
    do
        print '(/, "Current balance: ", F0.2)', balance
        print *, "1. Add Income"
        print *, "2. Add Expense"
        print *, "3. Show History"
        print *, "4. Exit"
        print *, "Enter choice: "
        
        ! Read input with error handling
        read(*,'(A)') input
        read(input, *, iostat=io) choice
        
        if (io /= 0) then
            print *, "Invalid input. Please enter 1-4"
            cycle
        end if

        select case (choice)
        case (1) ! Add Income
            print *, "Enter description: "
            read(*,'(A)') desc
            print *, "Enter amount: "
            read(*, *, iostat=io) amt
            if (io /= 0 .or. amt <= 0) then
                print *, "Invalid amount!"
                cycle
            end if
            
            if (transaction_count < max_trans) then
                transaction_count = transaction_count + 1
                transactions(transaction_count) = transaction(desc, amt, "Income")
                balance = balance + amt
            else
                print *, "Transaction limit reached!"
            end if

        case (2) ! Add Expense
            print *, "Enter description: "
            read(*,'(A)') desc
            print *, "Enter amount: "
            read(*, *, iostat=io) amt
            if (io /= 0 .or. amt <= 0) then
                print *, "Invalid amount!"
                cycle
            end if
            
            if (transaction_count < max_trans) then
                transaction_count = transaction_count + 1
                transactions(transaction_count) = transaction(desc, amt, "Expense")
                balance = balance - amt
            else
                print *, "Transaction limit reached!"
            end if

        case (3) ! Show History
            print '(/, "Transaction History:")'
            print *, "----------------------------------------------------"
            do i = 1, transaction_count
                print '(A10, ": ", A50, " - ", F0.2)', &
                    trim(transactions(i)%type), &
                    trim(transactions(i)%description), &
                    transactions(i)%amount
            end do
            print *, "----------------------------------------------------"

        case (4) ! Exit
            exit

        case default
            print *, "Invalid choice!"
        end select
    end do

    print *, "Goodbye!"
end program financial_tracker

File: KataFinancialTracker.php
<?php
class Transaction {
    public $description;
    public $amount;
    public $type;

    public function __construct($description, $amount, $type) {
        $this->description = $description;
        $this->amount = $amount;
        $this->type = $type;
    }
}

class FinancialTracker {
    const MAX_TRANS = 100;
    private $transactions = [];
    private $balance = 0.0;

    public function run() {
        echo "Simple Financial Tracker\n";
        echo "------------------------\n";

        while (true) {
            echo "\nCurrent balance: " . number_format($this->balance, 2) . "\n";
            echo "1. Add Income\n";
            echo "2. Add Expense\n";
            echo "3. Show History\n";
            echo "4. Exit\n";
            
            $choice = $this->readInput("Enter choice: ");
            
            switch ($choice) {
                case 1:
                    $this->addTransaction("Income");
                    break;
                case 2:
                    $this->addTransaction("Expense");
                    break;
                case 3:
                    $this->showHistory();
                    break;
                case 4:
                    echo "Goodbye!\n";
                    exit;
                default:
                    echo "Invalid choice!\n";
            }
        }
    }

    private function addTransaction($type) {
        if (count($this->transactions) >= self::MAX_TRANS) {
            echo "Transaction limit reached!\n";
            return;
        }

        $desc = $this->readInput("Enter description: ");
        $amt = $this->readNumericInput("Enter amount: ");

        if ($amt <= 0) {
            echo "Invalid amount!\n";
            return;
        }

        $this->transactions[] = new Transaction($desc, $amt, $type);
        $this->balance += ($type === "Income") ? $amt : -$amt;
    }

    private function showHistory() {
        echo "\nTransaction History:\n";
        echo str_repeat("-", 50) . "\n";
        
        foreach ($this->transactions as $transaction) {
            printf("%-10s: %-50s - %s\n",
                $transaction->type,
                $transaction->description,
                number_format($transaction->amount, 2)
            );
        }
        
        echo str_repeat("-", 50) . "\n";
    }

    private function readInput($prompt) {
        echo $prompt;
        return trim(fgets(STDIN));
    }

    private function readNumericInput($prompt) {
        while (true) {
            $input = $this->readInput($prompt);
            if (is_numeric($input)) {
                return (float)$input;
            }
            echo "Invalid input. Please enter a number.\n";
        }
    }
}

// Запуск приложения
$tracker = new FinancialTracker();
$tracker->run();
?>
